python/taichi/misc/gui.py

```diff
@@ -123,30 +123,53 @@ def line(self, begin, end, radius, color):
         self.canvas.path_single(begin[0], begin[1], end[0], end[1], color,
                                 radius)
 
-    def show(self, file=None):
+    def show(self, file_or_image=None):
```

Comment:
how about having two args: `file` and `image`? Then users can both show it and save it as files

Reply:
I would suggest them:
1. `gui.set_image(image); gui.show(file)`.
2. `gui.show(); ti.imwrite(image, file)`.
I like 2.

Reply:
Then i think this should just be `image`, because `imwrite` is already able to save to files? Again, if you think at the API layer, GUI is supposed to deal with one thing -- GUI. And we have separate APIs for file writings. WDYT?

Reply:
First of all, **this part does not belong to the event filtering system. Please revert**.

Please understand that having too many things in a PR makes reviewing really hard since the discussions cannot be focused on the real thing. This also increases the delay in merging your contribution since there are too many to discuss, and will result in nasty merge conflicts.


My thoughts on this specific issue: I agree that `ti.GUI.show` should not provide the option to write to disk. We might still need an API to get the image from the GUI or save the GUI image to disk, since sometimes people use drawing commands such as `ti.GUI.circles` to paint, and the painted result is only available from the GUI. Maybe we should do
 - `gui.set_image` to set a background image and allow people to paint on it
 - `gui.show(image=None)` to show the current canvas, or a plain 2D image (if `image` not none). Clear the temporary canvas after window content is updated.
 - `gui.screenshot(filename=None) -> np.ndarray` return the canvas. If `filename` is not `None`, use `imwrite` to save to disk. `ti.imwrite(gui.screenshot(), 'X.png')` sounds slightly verbose so I think a shortcut would be good.

The API design here should be put into a separate issue.